src/editor/action/action_label.cpp

```diff
@@ -88,4 +88,42 @@ void editor_action_label_delete::perform_without_undo(map_context& mc) const
 	mc.get_labels().set_label(loc_, "");
 }
 
+IMPLEMENT_ACTION(label_replace)
+
+editor_action* editor_action_label_replace::perform(map_context& mc) const
+{
+	editor_action_ptr undo(new editor_action_label_replace(new_loc_, loc_));
+
+	perform_without_undo(mc);
+	return undo.release();
+}
+
+void editor_action_label_replace::perform_without_undo(map_context& mc) const
+{
+	const terrain_label* old_label = mc.get_labels().get_label(loc_);
+
+	// Do nothing if label does not exist in original location
+	if (!old_label) {
+		return;
+	}
+
+	// Move label to new location by creating one in the new location
+	// with the same properties and clearing the old location
```

Comment:
This doesn't undo correctly.

* Create labels 'a' and 'b'
* Drag 'a' on top of 'b'
* Try to undo

The result is that 'a' moves back to where it originally was, but 'b' doesn't reappear.

Reply:
Thanks, you're right, this is something I completely overlooked. I was still having some trouble understanding how these editor actions worked, but I think I have a better understanding now. I've changed my approach to be simpler and similar to the original approach, but moving the logic to `drag_end_left` (when the drag is released).